### ISSUE TITLE
Nullability analysis not flowing through switches

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#35668]
**Version Used**: master

**Steps to Reproduce**:

Compile the following:

```csharp
using System;

#nullable enable

public class Person {
    public string? Name { get;}
}
public class C {
    public void M(Person p) {
        switch (p) {
            case { Name: var name } when name is null:
                Console.WriteLine("Hey anonymous");
            break;
            case { Name: var name }:
                Console.WriteLine($"Person has name of length { name.Length }!");
            break;
        }
    }
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACgqBiAOwFcMMJgM4C47X2qqcAzEQBMBAAqIoAezoEA3lQJKiQkngD8BAHIQwHOQQDmcGKQC+VC5UEiCAYXmLlNnABYCAWQAUEhNNkADgCUjpTK4QRQAO5oMADGABYEXsGhEekEcdD62rpwIAQAbhAIBNx6BGYEUQlcZXkEaFBlTBggThkZJACcXgBEABJwAJ4EEHQyw2BSDFB9QRRhncrACHAQANaLy8pZUDk6egXFpeUcZu1LO+E9XgAkfb7+BAnQ9RVSAGYE7HSGMEkDGcAHQAGS4/ySZgAhPNttdVusth0IlZwlYzEA=

**Expected Behavior**: no warning

**Actual Behavior**: warning CS8602: Dereference of a possibly null reference. `Console.WriteLine($"Person has name of length { name.Length }!");`

This looks related to https://github.com/dotnet/roslyn/issues/34828, but that has been fixed, and this still occurs in master on sharplab.

**Comment by jcouv** [comment]
Tagging @gafter to advise

**Comment by centreboard** [comment]
I assume this is related to the warning CS860 on `var length = line.Length;` in the following examples

``` csharp
public void Return(string? line) {
    switch (line) {
        case null:
            return;
    }
    var length = line.Length;
}

public void Continue() {
    while (true) {
        var line = Console.ReadLine();
        switch (line) {
            case null:
                continue;
        }
        var length = line.Length;
    }
}
```
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACgqBiAOwFcMMJgM4C47X2qqcAzEQBMBAMIEA3lQKyiQnABYCAJTgwGCOgAoSeAPwEMaOnACUUmXOtQA7mhgBjABYFtx0xemVrvuY+gORmYQKz9wogB2Ch9wgF8w3wA3CAQjLgBzGFcAXiMTOAA6ABlM7JjrBNjZRMEiZTEAezoYEwY4bS9E61tnTA5tGAR2ruqIghS0jw48prooRvZCtQgAE2KCzorx2TsHFzdp0Z2/AKggpgxQsZPfR2bWxjhtnarbyfS6LNz80xKy5wvXxvORVOJAA===

**Comment by jcouv** [comment]
In @YairHalberstadt's code snippet, the null test is in a `when` clause, which only affects that one switch branches but not other branches (by design).

The same code can be trivially [changed](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoUgYgDsBXDDCYDOAAjkoadNJwGZn9mABURQA9pWYBvUs1n8+OAgAYA/MwByEMC0nMA5nBgBuAL6kzJXvzzMAwlJlyrOACzMAsgAphCMRIAOAJQOJHJhzFAA7mgwAMYAFsyeQSHhacyx0DoaWnAgzDR0zCYgjunpigCcngBEABJwAJ7MEJTijWCi1FA1gUZl5czACHAQANb9oYOymVDZmtr5AG4QCAW5xaVT02FVngAkNT5+zPHQ69rMogBmzEyUejCJuuzaAHQAMmyPiSYAhL1JjshiNxkD0hYwhYTEA) (`case { Name: null }: ...`) to achieve the desired effect.

The code snippets from @centreboard do not use `when` clauses, so seem unrelated. Their current behavior seems correct to me. Please file a separate issue with actual vs. expected behavior.

I'll go ahead and close OP as by-design.
